Add support for semaphore file descriptor export

fayalalebrun opened the pull request:
Thanks to the `EXT_external_objects` OpenGL extension, there now exists a possibility to interoperate between OpenGL and Vulkan. Specifically, this extension allows OpenGL to use Vulkan textures and semaphores.

This PR takes a first step towards this, by allowing Vulkano to export its semaphores to Linux file descriptors, in the same way already possible with memory. Next, I will be working on allowing Vulkano to export its texture memory.

An example of semaphores being shared with Glium can be found here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7a9fa8702ca3f3b22813101daf861416

The related PR to add import functionality to Glium can be found here: https://github.com/glium/glium/pull/1939

    * Entries for Vulkano changelog:
        - Add creation of semaphores with exportable Linux file descriptor on
        - Add method to export file descriptor corresponding to semaphore

Eliah-Lakhin commented on vulkano/src/sync/semaphore.rs:
```diff
@@ -51,25 +55,47 @@ where
+    #[inline]
+    #[cfg(target_os = "linux")]
+    pub fn alloc_with_exportable_fd(device: D) -> Result<Semaphore<D>, OomError> {
+        Semaphore::alloc_impl(device, false, true)
```
The third parameter seem to be ignored in actual implementation.

Eliah-Lakhin commented on vulkano/src/sync/semaphore.rs:
```diff
@@ -51,25 +55,47 @@ where
+    /// Same as `alloc`, but allows exportable opaque file descriptor on Linux
+    #[inline]
+    #[cfg(target_os = "linux")]
+    pub fn alloc_with_exportable_fd(device: D) -> Result<Semaphore<D>, OomError> {
```
What was the reason to dedicate this function, and why it cannot be propagated to other platforms?

fayalalebrun replied:
I did this because most likely other handle types will have to be handled completely differently. This is a POSIX file descriptor so it won't work on Windows. You'd need an NT handle or one of the D3D specific types, and for the NT handle you also need to provide parameters, if I'm not mistaken.

Eliah-Lakhin reviewed (changes requested):
@fayalalebrun Thank you for your contribution. I left a couple of questions to be addressed.

gurchetansingh commented:
Why not use a builder?  Checkout the `DeviceMemoryBuilder`.

I think Vulkano should adopt a more standardized interface, so coding it up would be a non-brainer to anybody.  Right now, there's a bunch of styles.

Eliah-Lakhin commented:
> Why not use a builder?

@gurchetansingh I believe we could as long as it resolves the issues mentioned above.

Also, please notice that Linux test is failing. It requires an additional check in the test that the extension is available on particular device during the test running.

fayalalebrun commented:
@gurchetansingh I've added a `SemaphoreBuilder` and modeled it as close as possible to `DeviceMemoryBuilder`.

As part of this, I added some specific errors for Semaphores, but I had to convert them to the error type used in the `Swapchain` type and I'm not sure if I handled this the right way. I wrap `SemaphoreError` in `AcquisitionError` but they are both already wrapping `OomError`.

Eliah-Lakhin commented on vulkano/src/sync/semaphore/semaphore.rs:
```diff
@@ -0,0 +1,345 @@
+use std::ptr;
+use std::sync::Arc;
+
+use super::ExternalSemaphoreHandleType;
```
Better use full qualifier with `crate::...`

Eliah-Lakhin commented on vulkano/src/sync/semaphore/mod.rs:
```diff
@@ -0,0 +1,15 @@
+// notice may not be copied, modified, or distributed except
+// according to those terms.
+
+pub use self::external_semaphore_handle_type::ExternalSemaphoreHandleType;
```
Missing referenced type.

Eliah-Lakhin reviewed (changes requested):
@fayalalebrun Thank you for the fixes. The code looks fine at a first brief glance. But there seem to be missing file. I can't proceed with testing(and auto tests fail too).

fayalalebrun commented:
@Eliah-Lakhin 
There was a missing file, but I also broke one of the semaphore tests. I did not bother to check anything other than the semaphore export test, which was my mistake. With the latest commit, all tests should pass now.

Eliah-Lakhin commented on vulkano/src/sync/semaphore/semaphore.rs:
```diff
@@ -0,0 +1,345 @@
+where
+    D: SafeDeref<Target = Device>,
+{
+    pub fn new(device: D, must_put_in_pool: bool) -> Self {
```
@fayalalebrun As it was turned into a Builder, perhaps would be better to also turn `must_put_in_pool` parameter into a builder's function? Like `pub fn set_must_put_in_pool(mut self, flag: bool) -> Self` or something.

fayalalebrun replied:
Good idea. I've added it.

Eliah-Lakhin commented on vulkano/src/sync/semaphore/semaphore.rs:
```diff
@@ -0,0 +1,350 @@
+        }
+    }
+    /// Configures the semaphore to be added to the semaphore pool once it is destroyed.
+    pub(crate) fn in_pool(mut self) -> Self {
```
> pub(crate)

Is is hidden in public API by design?

fayalalebrun replied:
I believe so. The way semaphores behaved before there was no way to let users allocate semaphores which are meant for the pool. They were only created when needed. That's why I set it to only be available in the crate. Users can get semaphores from the pool using the `from_pool` method.

Eliah-Lakhin commented:
@fayalalebrun ok, all tests passed, I didn't detect any issues in my local tests too and the code looks fine. PR is merged :+1: Thank you very much for this improvement and for all the fixes!

pwaller commented on vulkano/src/sync/semaphore/semaphore.rs:
```diff
@@ -0,0 +1,355 @@
+        };
+
+        self.export_info = Some(export_info);
+        self.create.p_next = unsafe { std::mem::transmute(&export_info) };
```
This looks suspiciously similar to this other issue I raised:
https://github.com/vulkano-rs/vulkano/pull/1734#discussion_r771974391

Is the export_info referenced here living long enough? I'm unclear on that. I think this might need to refer to the `self.export_info.as_ref` or something similar.

fayalalebrun replied:
@pwaller Yes, I think there are two errors here. The first is that the struct the reference points to falls out of scope, since in this case the original is used instead of `self.export_info` (A copy is placed into the self struct). So the struct is destroyed when the scope of the function ends.

The second error is that a reference to the option is used instead of a reference to the underlying object. So `self.export_info.as_ref().unwrap()` can be used (We can unwrap here because we know the Option will always contain something at this point).

Could you make Github issues for these bugs? That way they are easy to keep track of. As comments to merged pull requests they might be easily forgotten.
